tls/s2n_prf.c

```diff
@@ -916,91 +999,68 @@ int s2n_prf_key_expansion(struct s2n_connection *conn)
+    POSIX_GUARD(s2n_hmac_init(
+            &conn->secure->server_record_mac,
+            conn->secure->cipher_suite->record_alg->hmac_alg,
+            key_material.server_mac.data,
+            key_material.server_mac.size));
 
     /* Make the client key */
     POSIX_GUARD(s2n_prf_make_client_key(conn, &key_material));
 
     /* Make the server key */
     POSIX_GUARD(s2n_prf_make_server_key(conn, &key_material));
 
-    /* Composite CBC does MAC inside the cipher, pass it the MAC key. 
+    /* Composite CBC does MAC inside the cipher, pass it the MAC key.
      * Must happen after setting encryption/decryption keys.
      */
-    if (conn->secure->cipher_suite->record_alg->cipher->type == S2N_COMPOSITE) {
-        POSIX_GUARD(conn->secure->cipher_suite->record_alg->cipher->io.comp.set_mac_write_key(&conn->secure->server_key, server_mac_write_key, mac_size));
-        POSIX_GUARD(conn->secure->cipher_suite->record_alg->cipher->io.comp.set_mac_write_key(&conn->secure->client_key, client_mac_write_key, mac_size));
-    }
-
-    /* TLS >= 1.1 has no implicit IVs for non AEAD ciphers */
-    if (conn->actual_protocol_version > S2N_TLS10 && conn->secure->cipher_suite->record_alg->cipher->type != S2N_AEAD) {
-        return 0;
-    }
-
-    uint32_t implicit_iv_size = 0;
-    switch (conn->secure->cipher_suite->record_alg->cipher->type) {
-        case S2N_AEAD:
-            implicit_iv_size = conn->secure->cipher_suite->record_alg->cipher->io.aead.fixed_iv_size;
-            break;
-        case S2N_CBC:
-            implicit_iv_size = conn->secure->cipher_suite->record_alg->cipher->io.cbc.block_size;
-            break;
-        case S2N_COMPOSITE:
-            implicit_iv_size = conn->secure->cipher_suite->record_alg->cipher->io.comp.block_size;
-            break;
-        /* No-op for stream ciphers */
-        default:
-            break;
+    if (cipher->type == S2N_COMPOSITE) {
+        POSIX_GUARD(cipher->io.comp.set_mac_write_key(&conn->secure->client_key, key_material.client_mac.data, key_material.client_mac.size));
+        POSIX_GUARD(cipher->io.comp.set_mac_write_key(&conn->secure->server_key, key_material.server_mac.data, key_material.server_mac.size));
     }
 
-    struct s2n_blob client_implicit_iv = { 0 };
-    POSIX_GUARD(s2n_blob_init(&client_implicit_iv, conn->secure->client_implicit_iv, implicit_iv_size));
-    struct s2n_blob server_implicit_iv = { 0 };
-    POSIX_GUARD(s2n_blob_init(&server_implicit_iv, conn->secure->server_implicit_iv, implicit_iv_size));
-    POSIX_GUARD(s2n_stuffer_read(&key_material, &client_implicit_iv));
-    POSIX_GUARD(s2n_stuffer_read(&key_material, &server_implicit_iv));
+    /* set IV if the size is non 0 */
+    POSIX_ENSURE_EQ(key_material.client_iv.size, key_material.server_iv.size);
+    POSIX_ENSURE_LTE(key_material.client_iv.size, S2N_TLS_MAX_IV_LEN);
+    POSIX_ENSURE_LTE(key_material.server_iv.size, S2N_TLS_MAX_IV_LEN);
```

Comment:
Nit: Technically you only need to check one of these, since you already ensured they were equal